File: medium/11.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    if (height.length <= 1) return 0;

    const getArea = (i, j) => Math.min(height[i], height[j]) * (j - i);
    let left = 0,
        right = height.length - 1,
        max = getArea(left, right);

    while (left < right) {
        if (height[left] > height[right])
            --right;
        else
            ++left;

        max = Math.max(max, getArea(left, right));
    }
    return max;
};


console.log(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]));
console.log(maxArea([1, 1]));

File: medium/11.cpp
/**
    # 11. Container With Most Water
*/
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0, r = height.size() - 1;
        int res = getArea(height, l, r);

        while(l < r){
            if(height[l] < height[r])
                ++l;
            else
                --r;
            res = max(res, getArea(height, l, r));    
        }

        return res;
    }

    int getArea(vector<int>& height, int l, int r){
        return (r - l) * min(height[l], height[r]);
    }
};
